### PR TITLE
BUG: Consistent timestamp fix for versions during publish recursive.

### PR DESCRIPTION
**Pull request by mfendeksilverstripe**
# Consistent timestamp fix for versions during publish recursive

## Problem

Calling` publishRecursive()` on a top level object like `Page` can produce inconsistent timestamps on version of nested objects. This breaks the version viewer and rollback functionality as they rely on correct timestamps of the versions.

## Example

* Page has 10 blocks
* Content author publishes the page which triggers recursive publish
* this operation creates published version of all related object starting from page going down to blocks
* let's assumes that operation started and ended with different timestamps (this may happen if the operations are slow or there are simply many objects to go through)
* the resulting versions of related objects will have different timestamps
* this can be emulated by adding `sleep()` to `onBeforeWrite()` calls

### Timestamps

Page - `2020-01-01 10:00:00`
Block 1 - `2020-01-01 10:00:00`
Block 2 - `2020-01-01 10:00:00`
Block 3 - `2020-01-01 10:00:00`
Block 4 - `2020-01-01 10:00:00`
Block 5 - `2020-01-01 10:00:00`
Block 6 - `2020-01-01 10:00:01`
Block 7 - `2020-01-01 10:00:01`
Block 8 - `2020-01-01 10:00:01`
Block 9 - `2020-01-01 10:00:02`
Block 10 - `2020-01-01 10:00:02`

### Why is this a problem

Version history viewer and rollback functionality rely on timestamps to retrieve the correct version. In this specific example the Page dictates the timestamps of the versions. This doesn't work however, because page was the first object to have the version written, so the timestamp used will exclude the other nested objects.

When viewing page version `2020-01-01 10:00:00` the blocks 1 through 5 will show correct content but blocks 6 through 10 will show outdated content as their version was written later.

Same applies to rollback.

## Solution

Lock the time at the start of the operation for the duration of the whole operation.

## Related issues

https://github.com/silverstripe/silverstripe-versioned/issues/289

## Notes

The diff is not very readable but there are no changes other than the callback wrap. You can view the [real diff here](https://github.com/silverstripe/silverstripe-versioned/pull/292/files?w=1).

## Dependencies

https://github.com/silverstripe/silverstripe-framework/pull/9630

### CODE REVIEW

**Comment by ScopeyNZ** [comment]
> The diff is not very readable but there are no changes other than the callback wrap.

Putting `&w=1` on the end of the PR URL makes it much more readable 😅

**Comment by mfendeksilverstripe** [comment]
This is amazing, thanks @ScopeyNZ

**Comment by michalkleiner** [comment]
It's the equivalent of [Hide whitespace changes](https://user-images.githubusercontent.com/233342/89355477-94412500-d70f-11ea-8aab-2ab3ab6542f1.png) setting. TIL.

**Comment by ScopeyNZ** [comment]
Ah yeah, the `?w=1` was around for a while before they added UI for it.

Back to this PR

I'm a little surprised this is actually relevant for rollbacks. Yes it does use timestamps but, although it's been a while since I looked at this code in-depth, I remember that it was relying on a timestamp on a record that was saved after related records. 

Maybe this is a new bug that came out of trying to prevent "unnecessary" saves? I vaguely remember seeing something like this in core.

File: src/RecursivePublishable.php [hunk 1]
```diff
@@ -57,41 +57,45 @@ class RecursivePublishable extends DataExtension
...
-        $owner = $this->owner;
+        $now = DBDatetime::now()->Rfc2822();
 
-        // get the last published version
-        $original = null;
-        if ($owner->hasExtension(Versioned::class) && $owner->isPublished()) {
-            $original = Versioned::get_by_stage($owner->baseClass(), Versioned::LIVE)
-                ->byID($owner->ID);
-        }
+        return DBDatetime::withMockNow($now, function () {
```
Comment (maxime-rainville) [inline src/RecursivePublishable.php:62]: I don't think this method exists.

Comment (tractorcow) [reply src/RecursivePublishable.php:62]: You could always create a `::withFixedNow()` method that automatically takes the current now(), and sets it as fixed for the duration. It'd look a lot more professional and less "should be a test-only method" :P

Comment (ScopeyNZ) [reply src/RecursivePublishable.php:62]: See https://github.com/silverstripe/silverstripe-framework/pull/9630

**Comment by ScopeyNZ** [comment]
I tested this pretty in depth a couple of years ago, and was only let down by the versions being created in the first place (see https://github.com/silverstripe/silverstripe-versioned/issues/195)

**Comment by mfendeksilverstripe** [comment]
@ScopeyNZ I reviewed the issue which you linked but I don't think it's related to my change set. The missing versions can be provided by using the [Versioned Snapshots](https://github.com/silverstripe/silverstripe-versioned-snapshots) module which lets you gain control of when versions are created.

**Comment by mfendeksilverstripe** [comment]
Pushed new changes - added unit tests & PR fixes.

File: src/RecursivePublishable.php (hunk 1 above)
Comment (maxime-rainville) [reply src/RecursivePublishable.php]: Duh ... just ignored me.

File: src/RecursivePublishable.php [hunk 2]
```diff
@@ -57,41 +57,45 @@ class RecursivePublishable extends DataExtension
...
-        $owner = $this->owner;
+        $now = DBDatetime::now()->Rfc2822();
 
-        // get the last published version
-        $original = null;
-        if ($owner->hasExtension(Versioned::class) && $owner->isPublished()) {
-            $original = Versioned::get_by_stage($owner->baseClass(), Versioned::LIVE)
-                ->byID($owner->ID);
-        }
+        return DBDatetime::withFixedNow($now, function () {
```
Comment (maxime-rainville) [inline src/RecursivePublishable.php:62]: The only problem is that `withFixedNow` will need to be in a new minor release. If you want to fix this in a patch release, you'll need to directly call `set_mock_now`.

Comment (mfendeksilverstripe) [reply src/RecursivePublishable.php:62]: I can't use `set_mock_now` as that approach will cause the previous mock value to be lost. Note that this functionality is used in many unit tests (already tried).

TLDR:

`clear_mock_now` sets the value to `null` not the previous value.

We will need the dependent PR to be merged first.

**Comment by mfendeksilverstripe** [comment]
@maxime-rainville Retargeted to `1`

**Comment by maxime-rainville** [comment]
Cool ... we'll just need to update the composer file to explicitly require framework `^4.7`.

**Comment by mfendeksilverstripe** [comment]
@maxime-rainville Right, done 👍

File: src/RecursivePublishable.php (hunk 2 above)
Comment (sminnee) [reply src/RecursivePublishable.php:62]: Dependent PR merged to `4`

**Comment by maxime-rainville** [comment]
Just restarted the build. Once its green, I'll have a go at running it locally as a final check to make sure it works.

File: tests/php/PublishRecursive/SlowDummyPage.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+use SilverStripe\Dev\TestOnly;
+
+/**
+ * Class SlowDummyPage
+ *
+ * @property int $NestedObjectID
+ * @method SlowDummyObject NestedObject()
+ * @package SilverStripe\Versioned\Tests\PublishRecursive
+ */
+class SlowDummyPage extends SiteTree implements TestOnly
```
Comment (maxime-rainville) [inline tests/php/PublishRecursive/SlowDummyPage.php:15]: CMS doesn't get installed by the build. So SiteTree doesn't exist in this context. Do you mind re-implementing your test with a Dummy DataObject instead?

Comment (mfendeksilverstripe) [reply tests/php/PublishRecursive/SlowDummyPage.php:15]: Yup, should be fine.

**Comment by mfendeksilverstripe** [comment]
@mfendeksilverstripe Pushed up unit test fixes.

File: tests/php/PublishRecursive/SlowDummyParent.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    /**
+     * @var array
+     */
+    private static $cascade_duplicates = [
+        'NestedObject',
+    ];
+
+    protected function onBeforeWrite()
+    {
+        sleep(2);
```
Comment (tractorcow) [inline tests/php/PublishRecursive/SlowDummyParent.php:60]: Adding sleeps to the unit tests is a very bad idea. I realise that this is time-critical, but this is going to only add incremental slowdowns to an already massive test suite.

A possible alternative; Clear mock now in your unit tests. Hook into `onBeforePublishRecursive` extension, and add a check to ensure that `DBDatetime` field has a specific mock assigned (i.e. fail if unmocked). You may need to use reflection to expose the protected static var.

If your code is behaving correctly, datetime will have a mock assigned during this process.

If the code is not behaving correctly, and the mock is unset (i.e. during onBeforePublishRecursive) then fail the test.

Comment (mfendeksilverstripe) [reply tests/php/PublishRecursive/SlowDummyParent.php:60]: Fair enough will try it out, thanks :)

**Comment by mfendeksilverstripe** [comment]
@maxime-rainville @tractorcow Pushed up sleep emulation (unit test fixes).

File: tests/php/PublishRecursive/SlowDummyParent.php [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        DBDatetime::set_mock_now($now);
+    }
+
+    /**
+     * @return DBDatetime|null
+     * @throws ReflectionException
+     */
+    private function getMockNow()
+    {
+        $propertyMockNow = new ReflectionProperty(DBDatetime::class, 'mock_now');
```
Comment (tractorcow) [inline tests/php/PublishRecursive/SlowDummyParent.php:96]: 👍

File: tests/php/PublishRecursive/SlowDummyParent.php [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        $this->emulateSleep(2);
+        parent::onBeforeWrite();
+    }
+
+    /**
+     * @param int $seconds
+     * @throws Exception
+     */
+    private function emulateSleep($seconds)
```
Comment (tractorcow) [inline tests/php/PublishRecursive/SlowDummyParent.php:73]: 👍

**Comment by maxime-rainville** [comment]
The Travis failure is caused by our old friend postgresql. I'm having a look to try to debug what's going on.

File: tests/php/PublishRecursive/PublishRecursiveTest.php [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        $parent = SlowDummyParent::create();
+        $parent->Title = 'Slow Parent';
+        $parent->NestedObjectID = (int) $object->ID;
+        $parent->write();
+
+        $parent->publishRecursive();
+
+        $versionsSufix = '_Versions';
+        $parentTable = SlowDummyParent::config()->get('table_name');
+        $parentVersionedTable = $parentTable . $versionsSufix;
```
Comment (maxime-rainville) [inline tests/php/PublishRecursive/PublishRecursiveTest.php:62]:
```suggestion
        $parentVersionedTable = '"' . $parentTable . $versionsSufix . '"';
```

File: tests/php/PublishRecursive/PublishRecursiveTest.php [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        $objectTable = SlowDummyObject::config()->get('table_name');
+        $objectVersionedTable = $objectTable . $versionsSufix;
```
Comment (maxime-rainville) [inline tests/php/PublishRecursive/PublishRecursiveTest.php:64]:
```suggestion
        $objectVersionedTable = '"' . $objectTable . $versionsSufix . '"';
```

File: tests/php/PublishRecursive/PublishRecursiveTest.php [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        $tables = [
+            $parentVersionedTable => $parent->ID,
+            $objectVersionedTable => $object->ID,
+        ];
+
+        $results = [];
+
+        foreach ($tables as $table => $id) {
+            $query = SQLSelect::create(
+                'LastEdited',
```
Comment (maxime-rainville) [inline tests/php/PublishRecursive/PublishRecursiveTest.php:74]:
```suggestion
                '"LastEdited"',
```

File: tests/php/PublishRecursive/PublishRecursiveTest.php [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                $table,
+                ['RecordID' => $id],
+                ['Version' => 'DESC'],
```
Comment (maxime-rainville) [inline tests/php/PublishRecursive/PublishRecursiveTest.php:77]:
```suggestion
                ['"RecordID"' => $id],
                ['"Version"' => 'DESC'],
```

**Comment by mfendeksilverstripe** [comment]
Thanks for finding the issue @maxime-rainville , I pushed up fixes and unit tests are now all green 👍
